### PR TITLE
PendingTX support

### PR DESCRIPTION
**Pull request by achamely**
This commit is dependent on the changes from https://github.com/mastercoin-MSC/omniEngine/pull/31

Adds support for inserting pending tx's into the db. 
Addresses #1066

### CODE REVIEW

**Comment by achamely** [comment]
Live on testnet.omniwallet.org  right now

**Comment by achamely** [comment]
Now live on testnet.omniwallet and staging.omniwallet

File: api/decode.py [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+@app.route('/<rawhex>')
+def decode_handler(rawhex):
+  return jsonify(decode(rawhex))
+
+
+def decode(rawhex):
+
+  transaction = decoderawtransaction(rawhex)['result']
+  rawtx = getrawtransaction(transaction['vin'][0]['txid'])
+  senders  = rawtx['result']['vout'][transaction['vin'][0]['vout']]['scriptPubKey']['addresses']
```
Comment (dexX7) [inline api/decode.py]: Care must be taken here: the sender is selected based on the combined values of all inputs, e.g.
Vin 0: sourceA, 0.5 BTC
Vin 1: sourceB, 0.4 BTC
Vin 2: sourceB, 0.2 BTC
... then the sender is sourceB.

But even more unfortunally: any other script type other than pay-to-pubkey-hash and pay-to-script-hash invalidates the whole transaction, if it's an input.

Related:
https://github.com/mastercoin-MSC/spec/issues/224 "Feature request: relax input and output requirements, P2SH"
https://github.com/mastercoin-MSC/spec/issues/225 "Easier transaction input selection"

**Comment by achamely** [comment]
Good point @dexX7 i'll try and rework the sender address selection to take that into account

**Comment by dexX7** [comment]
@achamely: or we could push the simplification on a protocol level.. ;)

File: api/decode.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  #get all multisigs
+  multisig_output = []
+  dest=""
+  for output in transaction['vout']:
+    if output['scriptPubKey']['type'] == 'multisig':
+      multisig_output.append(output) #grab msigs
+    else:
+      try:
+        for address in output['scriptPubKey']['addresses']:
+          if address not in (['1EXoDusjGwvnjZUyKkxZ4UHEf77z6A5S4P','mpexoDuSkGGqvqrkrjiFng38QPkJQVFyqv']+senders):
```
Comment (dexX7) [inline api/decode.py]: One more note: only the first sender is ignored as potential receiver *and* only pay-to-pubkey-hash or pay-to-script-hash outputs qualify as receiver, whereby other outputs are permitted. This is different to the input rules where any other script type as input invalidates the transaction.

[The rules below the encoding image starting with "a single transaction must be constructed satisfying the following requirements"](https://github.com/mastercoin-MSC/spec#class-b-transactions-also-known-as-the-multisig-method) cover everything, as far as I can see.

**Comment by achamely** [comment]
@dexX7 i'm working on this now (and did see those rules i'm adding in that logic).  but i found an interesting thing. what happens if (in your first example)  source A and source B have the same amount of input coins. It reads as that triggers an invalid tx, right?

**Comment by dexX7** [comment]
> what happens if (in your first example) source A and source B have the same amount of input coins. It reads as that triggers an invalid tx, right?

Great catch! Why do you assume it would invalidate the transaction?

**Comment by achamely** [comment]
Well its either invalid or we use the first / largest vin input.  But i couldn't find a distinction in the spec. (i'm on sococo if you want to take this realtime/easier)

**Comment by dexX7** [comment]
I'm not at home right now, but I'll check Master Core's behavior in the next hours. It won't invalidate, but on the first glimpse it's not necessarily the first, or the last vin, but might turn out to be something inbetween. Will report back. x_x

**Comment by achamely** [comment]
Ok, thanks

**Comment by achamely** [comment]
@dexX7 just heard from @m21 it looks like Core uses the first address in that situation.

**Pull request merged**